dlrover/trainer/tensorflow/executor/base_executor.py

```diff
@@ -0,0 +1,168 @@
+        self.task_type = task_type
+        self.task_id = task_id
+        self.role = task_type + ":" + str(task_id)
+        self.cluster_spec = TF_CONFIG["cluster"]
+        if self.task_type != "evaluator":
+            self.address = TF_CONFIG["cluster"][task_type][task_id]
+            # self.cluster_spec = TF_CONFIG["cluster"]
+
+            logger.info(
+                "cluster spec is {} \
+                    task_type is {} \
+                    task_id is {} \
+                    address is {}".format(
+                    self.cluster_spec,
+                    self.task_type,
+                    self.task_id,
+                    self.address,
+                )
+            )
+
+    def get_cluster_def(self, cluster_spec):
+        """
+        {
+             "ps": ["web04-pod2.default.svc:5002"],
+             "chief": ["web04-pod1.default.svc:5000"],
+             "worker": ["web04-pod2.default.svc:5000"],
+        }
+        """
+        mini_cluster_spec = {}
+        ps_hosts = []
+        worker_hosts = []
+        cluster_def = cluster_pb2.ClusterDef()
+        for job_name in cluster_spec:
+            if job_name == "ps":
+                job = cluster_def.job.add()
+                job.name = job_name
+                for task_index, address in enumerate(cluster_spec[job_name]):
+                    job.tasks[task_index] = address
+                    ps_hosts.append(address)
+            elif job_name == self.task_type:
+                job = cluster_def.job.add()
+                task_id = self.task_id
+                if job_name == "chief":
+                    job_name = "chief"
+                elif job_name == "worker":
+                    task_id = self.task_id + 1
+                job.name = job_name
+                job.tasks[task_id] = self.address
+        if self.task_type != "ps":
+            worker_hosts.append(self.address)
+        mini_cluster_spec["ps"] = ps_hosts
+        if self.task_type == "chief":
+            mini_cluster_spec["chief"] = worker_hosts
+        else:
+            mini_cluster_spec["worker"] = worker_hosts
+
+        if self.task_type == "chief":
+            print(1)
+            # mini_cluster_spec.pop("worker")
+            # mini_cluster_spec["chief"] = worker_hosts
+
+        self.mini_cluster_spec = mini_cluster_spec
+        logger.info("cluster def is:\n {}".format(cluster_def))
+        return cluster_def
+
+    def start_server(self):
+        if self.task_type != "evaluator":
+            logger.info("starting server")
+            self.server = server_lib.Server(
+                {"localhost": [self.address]}, protocol="grpc"
+            )
+            self.server.start()
+
+    def get_config(self, cluster_spec):
+        config = tf.estimator.RunConfig()
+        if self.task_type != "evaluator":
+            tf_config = os.environ["TF_CONFIG"]
+            tf_config = json.loads(tf_config)
+            tf_config["environment"] = "google"
```

Comment:
In tensorflow estimator/training.py, it starts server like this.
```
server = server_lib.Server(
        config.cluster_spec,
        job_name=config.task_type,
        task_index=config.task_id,
        config=session_config,
        start=False,
        protocol=config.protocol)
    server.start()
```
In the dynamic networking case, it would leads to conflicts
`// TODO(b/37868888): Remove this limitation when resolved
              status = errors::InvalidArgument(
                  "The ClusterSpec names the job and task index to be the same "
                  "names that were provided when the server booted. This is "
                  "currently not allowed. Job: ",
                  job.name(), ", task index: ", task.first);
              return;`

In order to enable dynamic networking, we will start tensorflow server like this to avoid conficts 
`def start_server(self):
        if self.task_type != "evaluator":
            logger.info("starting server")
            self.server = server_lib.Server(
                {"localhost": [self.address]}, protocol="grpc"
            )
            self.server.start()`

Reply:
Nice!